### ISSUE TITLE
[v6.1] General discussion

### ISSUE THREAD

**Issue by kisvegabor** [lvgl/lvgl#1132]
## Schedule
- Planned release date: 25 November 2019. 
- Feature stop: use `dev-7.0` for new features 

## Main new features

The most important goals are to
-  Moved to v7.0: Rework the examples repo: https://github.com/littlevgl/lv_examples/issues/31   Create tutorials, videos to gain more presence  

- Fix bugs and make v6.x more mature and stable 

#### Architecture  
- [x] Right-to-Left writing: #899 

#### Objects
- [x] Horizontal list: #1138 (implemented in #1140)
- [x] Add `lv_slider_set_sym`
- [x] Color picker (`lv_cpicker`) object types is added
- [x] Add `lv_lmeter_set_angle` #1262

#### Others
- [x] Sub pixel font rendering
- [x] asserts in API functions
- [x] Support compressed fonts
- [x] Use FontAwesome5 as symbols and add [new symbol](https://forum.littlevgl.com/t/glyph-additions/708) 
- Moved to v7.0: Store themes' data dynamically [LINK](https://forum.littlevgl.com/t/de-initing-themes/387/5)
- [x] Add `lv_printf` and `lv_label_set_text_fmt` for efficiently putting formatted text on labels
- [ ] Update the docs in `docs/dev-6.1`.

**Comment by embeddedt** [comment]
> Canvas blur/zoom

Are you planning to implement this by having users layer an `lv_blur`/`lv_zoom` object on top of the canvas?

**Comment by kisvegabor** [comment]
> Are you planning to implement this by having users layer an lv_blur/lv_zoom object on top of the canvas?

I was planning to do this similarly to rotate. So only for canvas. You draw something on the canvas and you can blur it or zoom into it.

**Comment by embeddedt** [comment]
@kisvegabor Okay. It would be neat to have the ability to blur other objects as well, but I can see the complexities of such an implementation due to the need for buffering all of the object's pixels.

**Comment by kisvegabor** [comment]
>  It would be neat to have the ability to blur other objects as well, but I can see the complexities of such an implementation due to the need for buffering all of the object's pixels.

It really would be nice but I suppose if you need these features you have some memory to buffer it. Anyway, I'll be thinking about it how to make these feature more generic and easier to use with other common objects. Maybe a function like `lv_canvas_draw_obj(canvas, obj, x, y)` could be handy to draw an object an all it's children to the canvas. 


I've just added an other nice canvas feature: texture mapping. It means the user gives an image and a 3D mesh and LittlevGL stretches the image to fit the mesh. It's enough to work with a 3D surface instead of a real object to avoid Z buffering. It can be used to add some perspective to objects or map a text a to arc, etc.

**Comment by embeddedt** [comment]
> I've just added an other nice canvas feature: texture mapping. It means the user gives an image and a 3D mesh and LittlevGL stretches the image to fit the mesh.

I'm afraid I didn't quite understand your description. Is there a simple code example so I can try it?

**Comment by kisvegabor** [comment]
Let's say you have an image with width and height = 100. and a polygon like `(0;0), (110, 0), (0,90), (120,130)`. Texture mapping should stretch the image to fit the polygon. With this, we could create distort the image to get a 3D-like effect.

**Comment by embeddedt** [comment]
@kisvegabor The idea sounds neat!

**Comment by puzrin** [comment]
I'm interested in more frequent releases. If sort ideas by value/cost ratio, i would suggest:

- 6.0.1 with font compression support

**Comment by embeddedt** [comment]
@puzrin Usually minor releases happen more frequently (i.e. 2-3 months) depending on the list of features to implement. We drained that list in 6.0 so we don't really have that much to do for 6.1 in comparison.

Most of the focus will be on writing more documentation and examples, which can be separate from the release itself.

**Comment by kisvegabor** [comment]
The last few months were a big hurry and very tough. So I'd like to slow down a little bit. 

For `v6.1` the most difficult features are:
- right to left support
- advanced masking
- texture mapping. 

The rest is quite straight forward. And takes only 1-2 days to implement. 

I'll create issues for all the planned features to see if somebody is interested in implementing them.

As now the main focus is on the examples, I'd say 3-4 month instead of 2-3 for the release.

**Comment by embeddedt** [comment]
> Maybe a function like `lv_canvas_draw_obj(canvas, obj, x, y)` could be handy to draw an object an all it's children to the canvas.

That would be neat! How would it work, exactly, since objects are currently associated with a display?

**Comment by kisvegabor** [comment]
> That would be neat! How would it work, exactly, since objects are currently associated with a display?

It's a nice opportunity which came with the multi-display support.
Have look at [lv_canvas_draw_rect](https://github.com/littlevgl/lvgl/blob/4a036d8c28725a4574ce3474bb960677ee3adde2/src/lv_objx/lv_canvas.c#L454).
It creates a dummy a display where the buffer is the canvas and draws the rectangle on it. In other words, it simulates the case when the library is refreshing the display but it's just dummy display now.
I haven't tested it with drawing real objects but it should work.

**Comment by embeddedt** [comment]
@kisvegabor Would it be possible to get input on the canvas with a similar hack (i.e. if I drew a button on the canvas, would I be able to click it)?

**Comment by kisvegabor** [comment]
Hm, interesting question. As usual, the canvas itself can receive events, however, as it doesn't keep track with the "things" drawn on the canvas can't search among them when a click happens.

**Comment by embeddedt** [comment]
@kisvegabor I've been thinking about this some more. What if we found some way of keeping the dummy display around for longer? Then a fake input device could also be attached to the canvas.

**Comment by kisvegabor** [comment]
How do you mean "Then a fake input device could also be attached to the canvas."?

If I understand the concept correctly, the problem is that on the canvas there are no objects, only pixels. So there is no way to tell if a clicked pixel belongs to button, label or anything else.

**Comment by aenchevich** [comment]
Hello!  :)
Hope this is the correct place to ask this question:
 I was using lvgl 5.x until recently but then the usual thing happened "help, problem! -- we won't fix in 5.x, use 6.x!"  so I started porting my code and .. omg are things different in 6! There's many new featrues, which is great, but also there are changes which break existing code even though it could be avoided.
 So... I would like to ask/suggest that whenever possible all 6.x changes are made as non-obtrusive as possible and with maximum 5.x backwards compatibility to make people's life easier (yes, that includes me :)) 

For example:
 There's a new param LV_ANIM_OFF/LV_ANIM_ON in many existing functions, e.g. a call like::
`lv_sw_on (switch2);`
now has to be:
`lv_sw_on (switch2, LV_ANIM_OF);`

So... the problems I see with this are:
a) existing code will not compile
b) people who do not want to use animation are forced to nevertheless include lv_anim in their code
c) as a direct consequence of b) the resulting binaries are bigger then necessary
d) the LV_USE_ANIM define becomes fuzzy: what does it control exactly? These new params are totally independent of it which is simply logically incorrect: 
  -- the LV_ANIM_ON makes no sense when LV_USE_ANIM is not defined and is confusing to see this in the code without any #ifdef guards, using #ifdef guards inside the function signature, n the other hand is ugly and makes the code hard to read. 

To fix the above I propose the following
- add funcs with signature like this (and guarded by USE_ANIM!):
```
#if defined ( LV_USE_ANIM )
    void lv_<OBJNAME>_use_anim( lv_obj * o, lv_anim_enable_t use_anim);
#endif

```
- then in your code, (if, and only if you want to use animation in your app!), call these funcs to enable animation for the objects when it is desired, e.g.:

```
#if defined ( LV_USE_ANIM )
#include "lv_anim.h"
#endif
  ...
#if defined ( LV_USE_ANIM )
    lv_sw_use_anim (switch2, LV_ANIM_ON);
#endif
    lv_sw_on (switch2);
```

This will solve all the problems I listed above:
 - existing code will work as before;
 - no need to worry about anim-related stuff in your code if you don't want it;
 - not passing dummy ANIM_OFF params around and no binary size increase;
 - the new feature will be controlled at compile time by the USE_ANIM define and at run time - by the LV_ANIM_ON/OFF values like it should be.

**Comment by kisvegabor** [comment]
@aenchevich We change major version (v5.x -> v6.x) when API breaking changes are introduced. On minor version (v6.0 -> v6.1) we keep backward compatibility. 

A major version change is really inconvenient but necessary when too many things are accumulated and it's necessary to change the architecture/API.

Your arguments are reasonable, but we can add them only in v7.0.

**Comment by embeddedt** [comment]
@kisvegabor What I meant is: right now in order to draw on the canvas a fake `lv_disp_t` is created just for the duration of that function. What if we provided a way to turn the canvas into a fully-functional display? That way objects could still be associated with it, and it could have input devices attached to it.

Essentially, the canvas would be a display inside a display.

**Comment by BrianPugh** [comment]
@kisvegabor @embeddedt now that v6.1 is underway, can we merge these 2 PRs into it? Thanks!

https://github.com/littlevgl/lvgl/pull/1119
https://github.com/littlevgl/lvgl/pull/1115

**Comment by kisvegabor** [comment]
@embeddedt 
> Essentially, the canvas would be a display inside a display.

I almost understand :slightly_smiling_face: 
- If the drawn objects can be rotated, zoomed, transformed etc, their coordinates will be changed so the input device module won't know where the object really is.
- If the drawn objects can't be rotated, etc, then how is it different from using a canvas as simple parent obejcts.

**Comment by kisvegabor** [comment]
@BrianPugh 
> now that v6.1 is underway, can we merge these 2 PRs into it? Thanks!

Sure, please send a new PR!

**Comment by aenchevich** [comment]
Gabor, 

I volunteer to add the `lv_<OBJNAME>_use_anim` funcs to all objects if you agree to this?

It's true that breaking some code between major versions is inevitable, and just as equally true is that it should be avoided if possible? Since it could be avoided in this particular case I think that this change would be a "Good Thing (tm)" 

And v7 is sooo far away...! 

:)

**Comment by kisvegabor** [comment]
@aenchevich 
Just to clarify for motivation:
- Would it'd be difficult (a lot of time) to migrate to v6.0 because of the changes in these prototypes? If so, please share your case. I can hardly imagine a UI where it would take more than half an hour.
- You don't use animations and it seems silly to type `LV_ANIM_OFF` every time. If so, the solution could be to change the prototype if the animations are disabled. (Not sure if it's really a good idea)

**Comment by embeddedt** [comment]
> If so, the solution could be to change the prototype if the animations are disabled. (Not sure if it's really a good idea)

I would suggest not adopting this approach. Changing the prototype depending on configuration makes it difficult to define a solid ABI (binary interface), whereas the current setup can just ignore the animation parameter if it's disabled, with no changes to the prototype.

It isn't really that much more work, in my opinion, to type `LV_ANIM_OFF`.

**Comment by kisvegabor** [comment]
> Changing the prototype depending on configuration makes it difficult to define a solid ABI (binary interface)

That's also true. It seems it really should remain for v7.0 :slightly_frowning_face:

**Comment by puzrin** [comment]
Not sure if it's for 6.1, but it really worth to solve scripting questions about all lvgl repos:

1. Recommended scripting solution for repos. Possibilities:
   - `make` - should be available everywhere, but not sure about windows (latest one should support linux CLI)
   - nodejs-based solutions like [Makefile.js](https://github.com/eslint/eslint/blob/master/Makefile.js)
2. Recommended testing approaches
   - My personal opinion is PIO https://docs.platformio.org/en/latest/ci/travis.html + may be nodejs's mocha for quick scripting https://github.com/mochajs/mocha (not C-related tasks).

I'm not sure about node.js and don't wish to push anyone about javascipt (i understand someone may not like JS at all). My opinion is based on fact, that new tools lv_i18n and lv_font_conv uses nodejs, and any intermediate user/dev will install nodejs anyway.

But i'll be ok with classic `make`/bash too, and ok to use PIO's built in python.

If you inspect lv_i18n - it uses make/nodejs, based on my personal opinion. Because i can't wait and need to select something. But it would be nice to use some unified (recommended?) approach for lvgl org. For example, i'm sure, authors of `lv_arduino` & `lv_binding_micropython` will be happy to automate some tasks to minimise human errors.

**Comment by embeddedt** [comment]
@puzrin Our goal is not to provide a centralized build solution. That makes it difficult to use LittlevGL in any project. One of the perks with LittlevGL is that you can use it by just copying it into a directory and compiling all of the source files.

**Comment by puzrin** [comment]
Let's separate maintainers an end users needs. Things i speak about are for maintainers. They need to automate updates and do integrity checks on lvgl changes.

**Comment by embeddedt** [comment]
> They need to automate updates

Updates are now almost entirely automated - the only manual steps are to write the release notes, update the PC simulator, and update the example projects.

> integrity checks on lvgl changes

There are plans to integrate CI at some point in the future, but at the moment, @kisvegabor has mentioned that he uses/tests LittlevGL nearly every day.

**Comment by kisvegabor** [comment]
We do have makefile based build system. This repo has [lvgl.mk](https://github.com/littlevgl/lvgl/blob/master/lvgl.mk) which is used by the [simulator](https://github.com/littlevgl/pc_simulator_sdl_eclipse/blob/master/Makefile) and the [linux frame buffer](https://github.com/littlevgl/lv_linux_frame_buffer/blob/master/Makefile) projects. `lv_examples` and `lv_drivers` also have their mk files.

**Comment by ShenRen** [comment]
When I work a Marlin UI project, I have port lvgl from C to C++ for convenience . 
Also ,I had  added some speciality to my code. like : smart pointer (detection lv_obj_t exist) , signal and slot (inspired by Qt) , and a improved lv_task . Those feature are easy to implement on C version lvgl.  I will post my code soon.
Last time , I had suggest a 3D feature which is in progress on v5.3 .

lvgl is not friendly for c++ class inherit the internal struct data . Because lv_obj_t `s memory allocate action is uncontrollable .   
At the first time , I try to use lv_obj_t as a class member . But convert a pointer from class to struct is verbose.
At the second time, I try to modified the lv_mem_allocate function and add new and delete function for class. Then all struct data can inherited by class . No more pointer conversion.

I hope , we can change the way lvgl create lv_obj_t .

**Comment by embeddedt** [comment]
Are you trying to put an `lv_obj_t` in a class, or an `lv_obj_t *` (pointer) in the class?

The first method won't work because LittlevGL needs to handle its own memory allocation.

I recommend you store the pointer inside your own custom class and use `lv_obj_create` and `lv_obj_del` to manage its creation/deletion.

**Comment by ShenRen** [comment]
@embeddedt  I put lv_obj_t * (pointer) in the class

First time :

```
class LVObject
{
    LV_OBJECT // memory func ( new and delete )`

    friend lv_res_t lvobjectSignalFunc (struct _lv_obj_t * obj, lv_signal_t sign, void * param);`
protected:
    lv_obj_t * m_this = nullptr; 
public:
.........
};
```

**Comment by ShenRen** [comment]
Second time:

```
/**
 * @brief lv_obj_t 
 */
class LVObjectNode
        : public lv_obj_t       
        , public LVLLNodeMeta  
{
protected:
    LVObjectNode() {}
};
```

```
/**
 * @brief The LVObject class
 * LVObject memory layout
 *                 LVObject                                    LVObjectNode
 *                 |                                           |   lv_obj_t
 *                 V                                           V   V
 * {LVMemoryHeader:[ (VTable) (LVObjectEnhance) {LVMemoryHeader:[ (lv_obj_t) (LVLLNodeMeta)  ] }  ] }
 */
class LVObject
        : public LVObjectEnhance 
        , public LVFakeMemHeader<LVObjectNode>  
        , public LVObjectNode   
{
    LV_MEMORY
    LV_NOCOPY(LVObject)

protected:
......
};
```

**Comment by amirgon** [comment]
@ShenRen What's wrong with your "First Time"?
Just keep a pointer to lv_obj_t as a class member, call the right `lv_<obj>_create` function in your constructor and `lv_obj_del` in the destructor.  
What's not working for you?

**Comment by ShenRen** [comment]
> @ShenRen What's wrong with your "First Time"?
> Just keep a pointer to lv_obj_t as a class member, call the right `lv_<obj>_create` function in your constructor and `lv_obj_del` in the destructor.
> What's not working for you?

It is work for me. 
I'm just saying that I have tried both ways.
I prefer the second method to the first.

**Comment by kisvegabor** [comment]
> Also ,I had added some speciality to my code. like : smart pointer (detection lv_obj_t exist) , signal and slot (inspired by Qt) , and a improved lv_task . Those feature are easy to implement on C version lvgl. I will post my code soon.

I'm curiously waiting to see your improvments!

**Comment by aenchevich** [comment]
> > If so, the solution could be to change the prototype if the animations are disabled. (Not sure if it's really a good idea)
> 
> I would suggest not adopting this approach. Changing the prototype depending on configuration makes it difficult to define a solid ABI (binary interface), whereas the current setup can just ignore the animation parameter if it's disabled, with no changes to the prototype.
> 
> It isn't really that much more work, in my opinion, to type `LV_ANIM_OFF`.

Regarding changing or not changing prototypes: actually it is my approach advocating NOT changing the prototype and keeping it as it was in 5.x. Adding new functions to implement new functionality is correct way and this is exactly what I am suggesting.

Regarding whether it is a lot of work: I started doing this. It is a lot of work (IMO). And given that it is unnecessary work and is breaking the interface I thought it would be a good idea to add the new function
- - -
(..and, yes typing LV_ANIM_OFF in all the functions even when not using it is kinda silly :D But really my goal is to keep the lib lean, mean  and clean - don't need something? don't waste bytes on it!)
(sorry I am responding slow - traveling a lot these days)

**Comment by kisvegabor** [comment]
> Regarding changing or not changing prototypes: actually it is my approach advocating NOT changing the prototype and keeping it as it was in 5.x. Adding new functions to implement new functionality is correct way and this is exactly what I am suggesting.

Okay, but it's already changed in v6.0 :smile: If we would change it after the release it wouldn't work for some people. We need to be strict in this regard else the whole versioning process wouldn't make sense.

**Comment by embeddedt** [comment]
> actually it is my approach advocating NOT changing the prototype and keeping it as it was in 5.x. Adding new functions to implement new functionality is correct way and this is exactly what I am suggesting.

That's what we do in minor releases. In major releases, usually some major change to the internals of the library has been performed.

Consider systems like Windows (or even Linux) which have been in production use for years. They have to keep a lot of compatibility logic around (even in Windows 10) to support applications designed for Windows XP.

That isn't suitable for an embedded system, in my opinion. In cases where internals are changed, eventually it makes more sense to change the prototype than to work around it's limitations.

We've been discussing animations. Here's an example of the difference between 5.x and 6.x:

5.x:

```c
/* Turn on a switch with animations */
lv_sw_anim_on(sw);
```

6.x:

```c
/* Turn on a switch with animations */
lv_sw_on(sw, LV_ANIM_ON);
```

I think we can agree that there isn't much difference between the two.

> don't need something? don't waste bytes on it!

I think the older 5.x approach actually takes up more code space, because (if you've felt/had the need to look at the disassembly) the compiler had to generate function prologue/epilogue for both functions. Also, I don't remember exactly what the code looked like, but it's possible that there was quite a bit of duplication between the two functions, which is even more code savings.

Another point: I don't know if this is true for every architecture, but on ARM, the compiler should only need to generate one more instruction for the extra parameters (2-4 bytes). That's a very small cost to pay for what (I think :man_shrugging: ) is a cleaner API.

> I started doing this. It is a lot of work (IMO).

Have you tried automating the process? If you use a regex you should be able to find all instances of `lv_sw_anim_on(*)` and replace them with `lv_sw_on(*, LV_ANIM_ON)`, regardless of the parameter name.

Let's look at another example: multi-display support. It just wouldn't be possible to cleanly keep the existing API in place without quite a bit of compatibility logic, which would take time to implement and probably wouldn't be worth the effort.

In summary, I understand that API breakage can be frustrating as a developer. But I think there comes a point when we have to drop old APIs to avoid code bloat.

**Comment by aenchevich** [comment]
@embeddedt 

Ok, firstly and most importantly: I understand Gabor's reason to say "ok but it is too late to do this in v6" and I am no longer advocating to add this to v6. 

Secondly (and this is really just a philosophical discussion at this point since we all agree that this will not go into v6) ... I do not agree with anything you say.

Your quoted comparison between 5 and 6 is not valid because it disregards the use of the LV_USE_ANIM define and that makes all of the subsequent reasoning invalid. 

The fact that a particular compiler is smart enough to optimize bad code is not a license to write bad code (to me anyway).

Yes I have tried to automate the process. Yes, if I use regex I can find all instances of LV_ANIM_ON in lw_sw. And then in lv_ddlist and lv_tabview, lv_bar, lv_win ... etc. 

I am not saying anything about multi-display support and I don't think it is related to this discussion. Nor am I advocating keeping existing interfaces at all costs. I am advocating keeping them whenever possible unless it is really necessary to change them. In this case it was possible to not break the interface, improve the scope of the USE_ANIM define and make peoples' lives easier. That was all I wanted when I proposed this change.

So we both hate code bloat. So I was wrong to say I do not agree with anything you say :)

**Comment by embeddedt** [comment]
>  am not saying anything about multi-display support and I don't think it is related to this discussion

So, for clarification, this discussion pertains purely to the animation prototypes? Fair enough.

> Your quoted comparison between 5 and 6 is not valid because it disregards the use of the LV_USE_ANIM define and that makes all of the subsequent reasoning invalid.

I didn't mention `LV_USE_ANIM` in that comparison, but I did have it in mind at the time.

In the event that `LV_USE_ANIM` is set to `0`, there is one unused parameter in that function call. That still looks clean in my opinion (a single API for turning on the switch, not two).

An unused parameter is not enough of a performance/code size drop to warrant having two functions (in my opinion).

> The fact that a particular compiler is smart enough to optimize bad code is not a license to write bad code

I wasn't referring to optimization with the 2-4 byte savings. I was just referring to the way most architectures work. Maybe I confused you.

> In this case it was possible to not break the interface, improve the scope of the USE_ANIM define and make peoples' lives easier.

I'll concede to that. There wasn't really an internal design choice that would have forced us to change it.

> So we both hate code bloat.

Correct. I do understand where you are coming from, but I also want to keep the LittlevGL API small, clean, and simple to use.

**Comment by ShenRen** [comment]
> > Also ,I had added some speciality to my code. like : smart pointer (detection lv_obj_t exist) , signal and slot (inspired by Qt) , and a improved lv_task . Those feature are easy to implement on C version lvgl. I will post my code soon.
> 
> I'm curiously waiting to see your improvments!

I have upload my project at [panda-3d-printer](https://github.com/panda-3d-printer). 
[LittlevGL_CPPPort](https://github.com/panda-3d-printer/LittlevGL_CPPPort) is a c++ version lvgl.
[PandaMini_MarlinUI](https://github.com/panda-3d-printer/PandaMini_MarlinUI) is a [Marlin](https://github.com/MarlinFirmware/Marlin) GUI , base on _LittlevGL_CPPPort_
[LittlevGL_QtSimulator](https://github.com/panda-3d-printer/LittlevGL_QtSimulator) is a LittlevGL Simulator which no need SDL. But it depends on Qt.

I haven't had time to sort out the project yet . There is only Chinese comment and no clear instructions.

For signal and slot , the c++ version code is here [lvsignalSlot](https://github.com/panda-3d-printer/LittlevGL_CPPPort/blob/master/core/lvsignalSlot.hpp). I only use  ```lv_ll``` to storage the connection between ```LVSignal``` and ```LVSlot```.

example for using:
```
    LVSignal IdleSignal;
    LVSlot idleSlot;
    idleSlot.setSlotFunc([](LVSignal* signal)
    {
        LV_LOG_INFO("idleSlot");
    });
    connect(&IdleSignal,&idleSlot,Connection::QueueConnect);// or DirectConnect
   ...
   //just call emit()
  IdleSignal.emit(); // then the slot function will be executed .
```

**Comment by kisvegabor** [comment]
@aenchevich 
Well, there wasn't any discussion when I've changed this API change so maybe I haven't considered all the possible problems. My reasons were the same what @embeddedt already mentioned:
- Avoid duplicating code internally. 
- Have fewer API functions.
- Make the API unchanged when configs are changing. It allows you to change the config without breaking your existing code.

**Comment by kisvegabor** [comment]
@ShenRen 
Could you open a topic for it in the [My projects](https://forum.littlevgl.com/c/my-projects) category in the forum?

**Comment by cron1c** [comment]
can you add "°" (degree) char to default font in the next release?

**Comment by embeddedt** [comment]
@cron1c That might be an extended Unicode character (beyond what basic ASCII supports by default). My guess is that it is deliberately missing from the default fonts in order to conserve flash space.

**Comment by kisvegabor** [comment]
> My guess is that it is deliberately missing from the default fonts in order to conserve flash space.

Yes, that's the case. The built-in font covers only the ASCII range and some symbols. If you need something else you can create a new font a disable the built-in one(s).

**Comment by kisvegabor** [comment]
I've update symbol fonts to use FontAwesome V5.0. Originally the symbols are split into three files but I merged them:  [FontAwesome5-Solid+Brands+Regular.woff.zip](https://github.com/littlevgl/lvgl/files/3634377/FontAwesome5-Solid%2BBrands%2BRegular.woff.zip)

I've added `LV_SYMBOL_BACKSPACE` and `LV_SYMBOL_PASTE`.

The new list of symbols to be used in the font converter:
```
61441, 61448, 61451, 61452, 61453, 61457, 61459, 61461, 61465, 61468, 
61473, 61478, 61479, 61480, 61502, 61508, 61512, 61515, 61516, 61517, 
61521, 61522, 61523, 61524, 61543, 61544, 61553, 61556, 61559, 61560, 
61561, 61563, 61587, 61589, 61636, 61637, 61639, 61671, 61674, 61683, 
61724, 61732, 61787, 61931, 62016, 62017, 62018, 62019, 62020, 62099, 
62189, 62810 
``` 

![image](https://user-images.githubusercontent.com/7599318/65302672-28e47800-db7c-11e9-9d12-2cfe42ed1096.png)

**Comment by kisvegabor** [comment]
The following symbols are added:
- EYE_OPEN
- EYE_CLOSE
- SD_CARD
- USB
- NEW_LINE

[FontAwesome5-Solid+Brands+Regular.woff.zip](https://github.com/littlevgl/lvgl/files/3651754/FontAwesome5-Solid%2BBrands%2BRegular.woff.zip)

```
      61441, 61448, 61451, 61452, 61452, 61453, 61457, 61459, 61461, 61465,
      61468, 61473, 61478, 61479, 61480, 61502, 61512, 61515, 61516, 61517,
      61521, 61522, 61523, 61524, 61543, 61544, 61550, 61552, 61553, 61556,
      61559, 61560, 61561, 61563, 61587, 61589, 61636, 61637, 61639, 61671,
      61674, 61683, 61724, 61732, 61787, 61931, 62016, 62017, 62018, 62019,
      62020, 62087, 62099, 62212, 62189, 62810, 63426, 63650
```

**Comment by char32** [comment]
Hi,
just wanted to ask: is there a plan to provide a C++ wrapper similar to the one @ShenRen created?
I know littlevGL is fully C++ compatible but having such a wrapper to use it completely object-orientated would be nice :-)
Thanks!

**Comment by embeddedt** [comment]
> I know littlevGL is fully C++ compatible but having such a wrapper to use it completely object-orientated would be nice :-)

It's not on the roadmap right now (as far as I know) but pull requests are welcome.

**Comment by kisvegabor** [comment]
> I know littlevGL is fully C++ compatible but having such a wrapper to use it completely object-orientated would be nice :-)

Would it be like a binding to C++? Why not? 
We already have a Micropython binding (created by @amirgon) which has a script to parse lvgl API and generate Python-C binding. Maybe that script can be used to generate C-C++ binding too.

**Comment by kisvegabor** [comment]
I've just pushed the new drawing engine and other new features to `dev-7.0`.

The new features should go to `dev-7.0`. WIP features will be merged to `dev-6.1` and it will be related in November 2019.

For the discussion of `dev-7.0` see https://github.com/littlevgl/lvgl/issues/1243

**Comment by kisvegabor** [comment]
I'm planning to release `6.1`  next Monday (25 November). Let me know if you think something is still missing.

**Comment by embeddedt** [comment]
I think all that's left is reviewing the documentation and making sure there is some basic information for all the new features.

I will try to write some documentation for the color picker this weekend.

**Comment by kisvegabor** [comment]
> I will try to write some documentation for the color picker this weekend.

That would be awesome!

**Comment by kisvegabor** [comment]
I've just created a `dev-6.1` branch in `docs`.

**Comment by kisvegabor** [comment]
I've just [released v6.1](https://github.com/littlevgl/lvgl/releases/tag/v6.1).

The docs still need to be polished a little bit but the main things are already there.

I've updated the Eclipse simulator project and I will update the STM32f429 and STM32769 project tomorrow.

**Comment by embeddedt** [comment]
I can update the STM32F746 project. Can it wait till this weekend or should I squeeze it in during the week?

**Comment by amirgon** [comment]
> I've just [released v6.1](https://github.com/littlevgl/lvgl/releases/tag/v6.1).

Updated `lv_binding_micropython` and `lv_micropython`.

**Comment by kisvegabor** [comment]
@embeddedt 
> Can it wait till this weekend or should I squeeze it in during the week?

It's absolutely fine at the weekend.

**Comment by kisvegabor** [comment]
I think our release policy needs to be improved. When a new version is released usually
1. I miss something, so the release is broken and a hotfix is required
2. Some people, who haven't used the `dev` branch, try the release and find issues. So other hotfixes are required.

What do you think what would be a better way to release new versions? For critical things now I simply change the release `tag` to HEAD but it's awkward...

Maybe make a `vx.y-rc` version on master and 1-2 week later make it `vx.y`?

**Comment by amirgon** [comment]
> Maybe make a `vx.y-rc` version on master and 1-2 week later make it `vx.y`?

rc is a good option.  
Another simpler option is a "code freeze" a week before the release. I noticed you changed a few things on dev-6.1 very close to the release.

Also, as we discussed in the past, CI with proper tests and various build configurations could be very helpful.

**Comment by kisvegabor** [comment]
> Another simpler option is a "code freeze" a week before the release.

I already use "feature freeze" but "code freeze" is really more powerful. It still requires people to use the dev branch so really `rc` seems a good way.

> Also, as we discussed in the past, CI with proper tests and various build configurations could be very helpful.

I'm also thinking about it. If I remember well, our last approach was to use `lv_micropython` to run the tests. However, it'd be great to see the test results in the open PRs of lvgl. But it requires to make possible to compile lvgl separately which is not the case today because it looks for `lv_conf.h`. With an `LV_NO_CONF` option we could create a "mode" which works only with `-D` defines and `lv_conf_checker.h`

**Comment by embeddedt** [comment]
> I think our release policy needs to be improved.

I noticed that we got a lot more pull requests over the last 12 hours than we usually get in a week. Perhaps we need a way of encouraging more people to use the `dev` branch.

CI is also a good option. We could also have an `lv_conf.h` specifically for the Travis build, but it's probably more stable to do it the way you're suggesting because that way there's one less file to update.

**Comment by puzrin** [comment]
@kisvegabor seems i have to merge dev branch of `lv_font_conv` and make new release?

**Comment by puzrin** [comment]
https://github.com/littlevghttps://github.com/littlevgl/lvgl/blob/master/library.json#L3 missed PIO version change. I don't know if postponed update ok or 6.1.1 needed.

**Comment by embeddedt** [comment]
There are a few hotfixes that were applied to 6.1 within 12 hours of being released :sweat_smile:, so I think we could just release "6.1.1" at the end of the week.

**Comment by kisvegabor** [comment]
@puzrin 
> seems i have to merge dev branch of lv_font_conv and make new release?

Yes, please.
@embeddedt Can you update the online converter by adding "Compressed" and "Subpixel" checkboxes? I think it's even more important than updating the projects.

> missed PIO version change. I don't know if postponed update ok or 6.1.1 needed.

Yes, let's do it in v6.1.1.

@embeddedt 
> We could also have an lv_conf.h specifically for the Travis build

I've no experience with Travis at all but I assume it'd be easier to make it work without external dependencies (such as `lv_conf.h` from another repo). To get started I think we would be fine to make simple build test with various configs and add tests later. 

> I think we could just release "6.1.1" at the end of the week.

I agree. Next time I'll make with an `rc` release first.

**Comment by puzrin** [comment]
FYI: released new 0.4.0 convertor, should be ok with 6.1.

**Comment by embeddedt** [comment]
> Can you update the online converter by adding "Compressed" and "Subpixel" checkboxes?

Done. https://github.com/littlevgl/lv_font_conv/commit/291354c5f8cf57a0ee3514bd7d192f1313a14c03

I'm updating the STM32F746 project now.

**Comment by embeddedt** [comment]
The STM32F746 SW4STM32 project has been updated. I will also update the Atollic one soon.

**Comment by kisvegabor** [comment]
> Can you update the online converter by adding "Compressed" and "Subpixel" checkboxes?
Done. littlevgl/lv_font_conv@291354c

Great, thank you! I have only one [question](https://github.com/littlevgl/lv_font_conv/commit/291354c5f8cf57a0ee3514bd7d192f1313a14c03#r36215810)

> The STM32F746 SW4STM32 project has been updated. I will also update the Atollic one soon.

Also great. I'll release v6.1.1 this week, but we can easily update to v6.1.1 as it contains only hotfixes, and no new features.

**Comment by kisvegabor** [comment]
@embeddedt thank you for updating the online font converter.
I've published it.

**Comment by embeddedt** [comment]
@kisvegabor [`library.json`](https://github.com/littlevgl/lvgl/blob/v6.1.1/library.json) still refers to `6.1.1`.

**Comment by kisvegabor** [comment]
Updated, thanks.

**Comment by embeddedt** [comment]
How about we open a new issue for discussing future release policies and close this one, since 6.1.1 has been released?

**Comment by kisvegabor** [comment]
> How about we open a new issue for discussing future release policies and close this one, since 6.1.1 has been released?

I agree. I'm opening it.


So I close this issue as v6.0 is released. 

Thank you very much for your contribution in v6.0!

**Comment by embeddedt** [comment]
@kisvegabor It might be a good idea for us to get into the habit of making an announcement shortly before support for old releases is dropped. 7.0 was released on May 19th, 2020, so following the 1-year-after-major-release policy, we should be able to drop official support for 6.1 after May 19th, 2021.

Personally, I think we may as well drop support at the same time that v8 releases, which is looking to be late May or early June.

I'm wondering if it would be a good idea to make an announcement on the forum and pin a GitHub issue, just so that everyone's aware of this. The 5.x deprecation wasn't as clearly delineated, and with the increasing popularity of LVGL, I feel that this is probably good to do as a courtesy.

**Comment by kisvegabor** [comment]
I definitely agree to clearly communicate it. But instead of a pinned issue, I think we can add it to the changelog (e.g. to the last release of a major version),  or int the beginning of the change log, and/or in a table in the docs.
